### PR TITLE
GH-34035: [C++] Internal header file included from public one breaks build of external projects

### LINKED ISSUE #34035: [C++] Internal header file included from public one breaks build of external projects
### Describe the bug, including details regarding any error messages, version, and platform.

There are a number of public header files that include internal header files:
```
$ find cpp/src/arrow/ -name "*.h" | xargs grep -l '_internal.h' | grep -v '_internal.h$'
cpp/src/arrow/compute/exec/swiss_join.h
cpp/src/arrow/compute/exec/util.h
cpp/src/arrow/compute/kernels/common.h
cpp/src/arrow/compute/kernels/row_encoder.h
cpp/src/arrow/dataset/test_util.h
```
This causes the build of external projects to break when including the public header because the internal one cannot be found since it is not part of the install.

Besides fixing the above public headers, it would be useful to add a (push-time?) check that ensures the above list is empty.

### Component(s)

C++

### PR DESCRIPTION
**Pull request by rtpsw**
See #34035
* Closes: #34035

### CODE REVIEW

File: cpp/src/arrow/compute/exec/util.h [hunk 1]
```diff
@@ -57,6 +56,15 @@
...
 namespace arrow {
+
+namespace internal {
+namespace tracing {
+
+struct Scope;
+
+}  // namespace tracing
+}  // namespace internal
+
```
Comment (westonpace) [inline cpp/src/arrow/compute/exec/util.h:67]: Can we instead add `Scope` to `arrow/util/type_fwd.h` (and include that here)

File: cpp/src/arrow/compute/exec/util.h [hunk 2]
```diff
@@ -234,6 +242,8 @@ class bit_util {
 }  // namespace util
 namespace compute {
 
+class ExecNode;
```
Comment (westonpace) [inline cpp/src/arrow/compute/exec/util.h:245]: Instead can we include `arrow/compute/type_fwd.h`?

File: cpp/src/arrow/compute/exec/util.h [hunk 3]
```diff
@@ -36,7 +36,6 @@
 #include "arrow/util/logging.h"
 #include "arrow/util/mutex.h"
 #include "arrow/util/thread_pool.h"
-#include "arrow/util/tracing_internal.h"
```
Comment (westonpace) [inline cpp/src/arrow/compute/exec/util.h:39]: Note, this header is still an internal header as it includes `arrow/util/logging.h` which is perhaps our most egregious improperly named header.

**Review by westonpace — CHANGES_REQUESTED** [review]
Thanks for taking this on.

File: cpp/src/arrow/compute/exec/util.h (hunk 3 above)
Comment (rtpsw) [reply cpp/src/arrow/compute/exec/util.h:39]: Though this won't break an external project build, right?

File: cpp/src/arrow/compute/exec/util.h (hunk 1 above)
Comment (rtpsw) [reply cpp/src/arrow/compute/exec/util.h:67]: Handling this in the next commit.

File: cpp/src/arrow/compute/exec/util.h (hunk 2 above)
Comment (rtpsw) [reply cpp/src/arrow/compute/exec/util.h:245]: Looks like the header is already included and the forward declaration here can be removed.

File: cpp/src/arrow/compute/exec/util.h (hunk 3 above)
Comment (westonpace) [reply cpp/src/arrow/compute/exec/util.h:39]: Ah, I don't know.  But we have a check that ensures we don't expose `logging.h` from `api.h`.  So we need to keep `map_node.h` out of `api.h` until this is resolved.

Comment (rtpsw) [reply cpp/src/arrow/compute/exec/util.h:39]: Done.

**Pull request merged**